File: src/main/java/de/chaos/swlnmngr/UpdateChecker.java
package de.chaos.swlnmngr;

import de.chaos.swlnmngr.config.Config;
import lombok.experimental.UtilityClass;
import org.apache.http.client.methods.CloseableHttpResponse;
import org.apache.http.client.methods.HttpGet;
import org.apache.http.impl.client.CloseableHttpClient;
import org.apache.http.impl.client.HttpClients;
import org.json.JSONArray;
import org.json.JSONObject;
import org.json.JSONTokener;

import java.io.IOException;
import java.net.URI;

@UtilityClass
public class UpdateChecker {

    private static final URI repoUrl = URI.create("https://steamwar.de/devlabs/api/v1/repos/Chaoscaot/SteamwarLinkManager/releases?draft=false&pre-release=" + Config.PRE_RELEASES + "&page=1&limit=1");

    public static final String CURRENT_VERSION;

    static {
        try {
            CURRENT_VERSION = new String(UpdateChecker.class.getResourceAsStream("/jar.version").readAllBytes());
        } catch (IOException e) {
            Main.getLogger().error("Could not read Version", e);
            System.exit(1);
            throw new SecurityException(e);
        }
    }

    public static void checkForUpdates() {
        try (CloseableHttpClient client = HttpClients.createMinimal()) {
            HttpGet get = new HttpGet(repoUrl);
            try (CloseableHttpResponse response = client.execute(get)) {
                JSONArray array = new JSONArray(new JSONTokener(response.getEntity().getContent()));
                JSONObject latestObject = array.getJSONObject(0);
                String latestVersion = latestObject.getString("tag_name");
                if(!latestVersion.equals(CURRENT_VERSION)) {
                    Main.getLogger().info("The Running Jar is not the Latest release Version\n\tYour Version: {}\n\tLatest Release Version: {}", CURRENT_VERSION, latestVersion);
                    Main.getLogger().info("Download the Latest Jar here: https://steamwar.de/devlabs/Chaoscaot/SteamwarLinkManager/releases");
                }
            }
        } catch (IOException e) {
            Main.getLogger().error("Could not fetch Updates", e);
        }
    }
}

File: src/main/java/de/chaos/swlnmngr/UpdateChecker.kt
package de.chaos.swlnmngr

import de.chaos.swlnmngr.config.CLIConfig
import de.chaos.swlnmngr.config.cliConfig
import de.chaos.swlnmngr.route.routes.Route
import org.apache.http.client.methods.HttpGet
import org.apache.http.impl.client.HttpClients
import java.io.File
import java.io.IOException
import java.net.URI
import java.net.URL
import kotlin.system.exitProcess
import java.nio.file.Files
import java.nio.file.StandardOpenOption
import kotlin.math.log

private val repoURL :URI = URI.create("https://github.com/Chaoscaot/SteamwarLinkManager/releases/latest")
private val download = URI.create("$repoURL/download/SteamWarLinkManager.jar")

private val CURRENT_VERSION :String by lazy {
    try {
        return@lazy String(Route::class.java.getResourceAsStream("/jar.version")!!.readAllBytes())
    } catch (e :IOException) {
        logger.error("Could not read Version", e)
        exitProcess(1)
    }
}

data class UpdateInfo(val updateAvailable: Boolean, val version: String)

fun updateAvailable(): UpdateInfo {
    val httpClient = HttpClients.custom().disableCookieManagement().disableRedirectHandling().build()
    val httpGet = HttpGet(repoURL)
    val response = httpClient.execute(httpGet)
    if(cliConfig.isDebug) {
        logger.debug("Response Code: ${response.statusLine.statusCode}")
        response.allHeaders.forEach {
            logger.debug("${it.name}: ${it.value}")
        }
    }
    val latestVersion = response.getFirstHeader("location").value.split("/").last()
    response.close()
    httpClient.close()
    return UpdateInfo(latestVersion != CURRENT_VERSION, latestVersion)
}

fun checkForUpdates() {
    val updateInfo = updateAvailable()
    if (updateInfo.updateAvailable) {
        logger.info("The Running Jar is not the Latest release Version\n\tYour Version: {}\n\tLatest Release Version: {}",
                CURRENT_VERSION, updateInfo.version)
        logger.info("Run the Jar with `updatejar` to update to the latest Version")
    }
}

fun downloadToFile(): File {
    Files.newOutputStream(File(cliConfig.installDir, "SteamWarLinkManager-new.jar").toPath(), StandardOpenOption.CREATE_NEW, StandardOpenOption.TRUNCATE_EXISTING).use {
        URL(download.toString()).openStream().transferTo(it)
        it.flush()
    }
    return File(cliConfig.installDir, "SteamWarLinkManager-new.jar")
}
